relay-cogs/src/lib.rs

```diff
@@ -0,0 +1,257 @@
+    MetricsTransactions,
+    /// Metrics in the spans namespace.
+    MetricsSpans,
+    /// Metrics in the sessions namespace.
+    MetricsSessions,
+    /// Metrics in the custom namespace.
+    MetricsCustom,
+    /// Metrics in the unsupported namespace.
+    ///
+    /// This is usually not emitted, since metrics in the unsupported
+    /// namespace should be dropped before any processing occurs.
+    MetricsUnsupported,
+}
+
+impl AppFeature {
+    /// Returns the string representation for this app feature.
+    pub fn as_str(&self) -> &'static str {
+        match self {
+            Self::Unattributed => "unattributed",
+            Self::UnattributedMetrics => "unattributed_metrics",
+            Self::UnattributedEnvelope => "unattributed_envelope",
+            Self::Transactions => "transactions",
+            Self::Errors => "errors",
+            Self::Spans => "spans",
+            Self::Sessions => "sessions",
+            Self::ClientReports => "client_reports",
+            Self::CheckIns => "check_ins",
+            Self::Replays => "replays",
+            Self::MetricMeta => "metric_meta",
+            Self::MetricsTransactions => "metrics_transactions",
+            Self::MetricsSpans => "metrics_spans",
+            Self::MetricsSessions => "metrics_sessions",
+            Self::MetricsCustom => "metrics_cusomt",
+            Self::MetricsUnsupported => "metrics_unsupported",
+        }
+    }
+}
+
+impl utils::Enum<16> for AppFeature {
+    fn from_index(index: usize) -> Option<Self> {
+        let r = match index {
+            0 => Self::Unattributed,
+            1 => Self::UnattributedMetrics,
+            2 => Self::UnattributedEnvelope,
+            3 => Self::Transactions,
+            4 => Self::Errors,
+            5 => Self::Spans,
+            6 => Self::Sessions,
+            7 => Self::ClientReports,
+            8 => Self::CheckIns,
+            9 => Self::Replays,
+            10 => Self::MetricMeta,
+            11 => Self::MetricsTransactions,
+            12 => Self::MetricsSpans,
+            13 => Self::MetricsSessions,
+            14 => Self::MetricsCustom,
+            15 => Self::MetricsUnsupported,
+            _ => return None,
+        };
+
+        Some(r)
+    }
+
+    fn to_index(value: Self) -> usize {
+        match value {
+            Self::Unattributed => 0,
+            Self::UnattributedMetrics => 1,
+            Self::UnattributedEnvelope => 2,
+            Self::Transactions => 3,
+            Self::Errors => 4,
+            Self::Spans => 5,
+            Self::Sessions => 6,
+            Self::ClientReports => 7,
+            Self::CheckIns => 8,
+            Self::Replays => 9,
+            Self::MetricMeta => 10,
+            Self::MetricsTransactions => 11,
+            Self::MetricsSpans => 12,
+            Self::MetricsSessions => 13,
+            Self::MetricsCustom => 14,
+            Self::MetricsUnsupported => 15,
+        }
+    }
+}
```

Comment:
Given that we need to specify the indices + the enum size manually, why not go with

```rust
enum AppFeature {
    a = 0,
    b = 1,
    // ...
}

struct AppFeatures([NonZeroUsize; 16]);

impl AppFeatures {
    fn get(&self, index: AppFeature) -> Option<&NonZeroUsize> {
        self.0.get(index as usize)
    }
}
```

Would save a lot of LoC, plus the compiler even ensures that you do not use the same discriminant twice.

Reply:
That is already possible with the current `AppFeature`, it implicitly starts at 0 (I'll actually update `to_index` to use the cast). 

But we still need to convert back from an index to the enum variant, which comes down to the same `from_index`. I initially had all of this inlined but it just became unreadable and I preferred having the array logic separate from the `AppFeatures`.

Reply:
How about we just use a `BTreeMap` then? Do we have reason to believe that those few allocations will have a measurable effect on performance?